pkg/rescaffold/migrate.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+	http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rescaffold
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+
+	"github.com/spf13/afero"
+	"sigs.k8s.io/kubebuilder/v3/pkg/config/store"
+	"sigs.k8s.io/kubebuilder/v3/pkg/config/store/yaml"
+	"sigs.k8s.io/kubebuilder/v3/pkg/machinery"
+	"sigs.k8s.io/kubebuilder/v3/pkg/plugin/util"
+)
+
+type MigrateOptions struct {
+	InputDir  string
+	OutputDir string
+}
+
+const DefaultOutputDir = "output-dir"
+
+func (opts *MigrateOptions) Rescaffold() error {
+	store := yaml.New(machinery.Filesystem{FS: afero.NewOsFs()})
+	if err := store.LoadFrom(opts.InputDir); err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+	// create output directory
+	if err := os.MkdirAll(opts.OutputDir, os.ModePerm); err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+	// use the new directory to set up the new project
+	if err := os.Chdir(opts.OutputDir); err != nil {
+		log.Fatal(err)
+	}
+	// init project with plugins
+	if err := kubebuilderInit(store); err != nil {
+		log.Fatal(err)
+	}
+	return nil
+}
+
+func (opts *MigrateOptions) Validate() error {
+	cwd, err := os.Getwd()
+	if err != nil {
+		log.Fatal(err)
+	}
+	// get PROJECT path from command args
+	inputPath, err := getInputPath(cwd, opts.InputDir)
+	if err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+	opts.InputDir = inputPath
+	// get output path from command args
+	opts.OutputDir = getOutputPath(cwd, opts.OutputDir)
+	// check whether the kubebuilder binary is accessible
+	_, err = exec.LookPath("kubebuilder")
+	return err
+}
+
+func getInputPath(cwd string, inputPath string) (string, error) {
+	if inputPath == "" {
+		inputPath = cwd
+	}
+	projectPath := fmt.Sprintf("%s/%s", inputPath, yaml.DefaultPath)
+	if _, err := os.Stat(projectPath); os.IsNotExist(err) {
+		return "", fmt.Errorf("PROJECT path: %s does not exist. %v", projectPath, err)
+	}
+	return projectPath, nil
+}
+
+func getOutputPath(cwd, outputPath string) string {
+	if outputPath == DefaultOutputDir {
+		outputPath = fmt.Sprintf("%s/%s", cwd, outputPath)
+	}
+	return outputPath
+}
+
+func kubebuilderInit(store store.Store) error {
+	var args []string
+	args = append(args, "init")
+	return util.RunCmd("kubebuilder init", "kubebuilder", args...)
```

Comment:
Is this ready in use OR we will have future development for this?
(In other words, do we need to add a todo comment here?)

Reply:
Also, will we call Kubebuilder executable to re-scaffold the project in the future? It feels hacky to me...

Reply:
Yes, this is in use. We will call that to init the project from 'PROJECT file'